src/bank.rs

```diff
@@ -741,6 +741,30 @@ mod tests {
         assert_eq!(bank.get_balance(&mint.pubkey()), 1);
     }
 
+    // Test how fees are dealt in contracts that error out.
+    // This test verifies current incorrect behavior. Contracts that cause
+    // execution on chain should pay the fee.
+    #[test]
+    fn test_failed_with_fee() {
+        let mint = Mint::new(1);
+        let bank = Bank::new(&mint);
+        let dest = Keypair::new();
+
+        // source with 0 contract context
+        let tx = Transaction::system_move(&mint.keypair(), dest.pubkey(), 1, mint.last_id(), 1);
+        let signature = tx.signature;
+        assert!(!bank.has_signature(&signature));
+        let res = bank.process_transaction(&tx);
+        // This is the potentially wrong behavior
+        // result failed, but signature is registered
```

Comment:
This is the behaviour I was seeing from the SDK side yesterday too.  

Caller really needs to know if the executed transaction actually succeeded or not.  A transaction-specific result code that can be fetched via the RPC API use the transaction `signature`

Reply:
Change this from a `HashSet<Signature>` to a `HashMap<Signature, Result<()>>`:

https://github.com/solana-labs/solana/blob/master/src/bank.rs#L126

Reply:
yea, i agree.  last_sigs table can be modified to store a HashMap<Signature,ErrorCode> . something cheap that fits into a u64.